File: JavaScript/DSA/LinkedList.js
// A Linked List is a linear data structure that consists of a sequence of nodes, where each node contains a value and a reference to the next node in the sequence

class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}

class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }

  // add a node to the end of the list
  append(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = this.head;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
    this.length++;
  }

  // add a node to the beginning of the list
  prepend(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = this.head;
    } else {
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
  }

  // remove the first occurrence of a node with the specified value
  remove(value) {
    let currentNode = this.head;
    let previousNode = null;
    while (currentNode !== null) {
      if (currentNode.value === value) {
        if (previousNode === null) {
          this.head = currentNode.next;
        } else {
          previousNode.next = currentNode.next;
        }
        this.length--;
        return;
      }
      previousNode = currentNode;
      currentNode = currentNode.next;
    }
  }

  // get the node at the specified index
  getNodeAtIndex(index) {
    if (index < 0 || index >= this.length) {
      return null;
    }
    let currentNode = this.head;
    for (let i = 0; i < index; i++) {
      currentNode = currentNode.next;
    }
    return currentNode;
  }

  // print the values of all nodes in the list
  printList() {
    let currentNode = this.head;
    while (currentNode !== null) {
      console.log(currentNode.value);
      currentNode = currentNode.next;
    }
  }
}

// usage

const myList = new LinkedList();

myList.append(10);
myList.append(20);
myList.append(30);
myList.prepend(5);

console.log(myList.length); // Output: 4

myList.printList(); // Output: 5 10 20 30

myList.remove(20);

console.log(myList.length); // Output: 3

myList.printList(); // Output: 5 10 30

console.log(myList.getNodeAtIndex(2).value); // Output: 30

File: C&C++/DSA/LinkedList.cpp
/*
A linked list is a dynamic data structure in C++ that consists of a sequence of nodes. Each node contains data and a pointer to the next node in the sequence. The first node in the sequence is called the head, and the last node points to NULL.
*/

#include <iostream>

using namespace std;

// Define a node structure
struct Node
{
    int data;
    Node *next;
};

// Define the linked list class
class LinkedList
{
private:
    Node *head;

public:
    // Constructor
    LinkedList()
    {
        head = NULL;
    }

    // Add a node to the end of the list
    void append(int value)
    {
        Node *new_node = new Node;
        new_node->data = value;
        new_node->next = NULL;

        if (head == NULL)
        {
            head = new_node;
        }
        else
        {
            Node *temp = head;
            while (temp->next != NULL)
            {
                temp = temp->next;
            }
            temp->next = new_node;
        }
    }

    // Insert a node at the beginning of the list
    void prepend(int value)
    {
        Node *new_node = new Node;
        new_node->data = value;
        new_node->next = head;
        head = new_node;
    }

    // Delete the first occurrence of a value in the list
    void remove(int value)
    {
        if (head == NULL)
        {
            return;
        }

        if (head->data == value)
        {
            Node *temp = head;
            head = head->next;
            delete temp;
            return;
        }

        Node *temp = head;
        while (temp->next != NULL)
        {
            if (temp->next->data == value)
            {
                Node *temp2 = temp->next;
                temp->next = temp2->next;
                delete temp2;
                return;
            }
            temp = temp->next;
        }
    }

    // Get the length of the list
    int length()
    {
        int count = 0;
        Node *temp = head;
        while (temp != NULL)
        {
            count++;
            temp = temp->next;
        }
        return count;
    }

    // Print the values in the list
    void print()
    {
        Node *temp = head;
        while (temp != NULL)
        {
            cout << temp->data << " ";
            temp = temp->next;
        }
        cout << endl;
    }
};

int main()
{
    LinkedList list;

    list.append(5);
    list.append(3);
    list.append(7);

    list.print(); // Output: 5 3 7

    list.prepend(2);

    list.print(); // Output: 2 5 3 7

    list.remove(3);

    list.print(); // Output: 2 5 7

    cout << "Length: " << list.length() << endl; // Output: Length: 3

    return 0;
}
